ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiSnapshotWorldStateKeyValueStorage.java

```diff
@@ -0,0 +1,120 @@
+import org.hyperledger.besu.ethereum.trie.StoredNodeFactory;
+import org.hyperledger.besu.ethereum.worldstate.StateTrieAccountValue;
+import org.hyperledger.besu.ethereum.worldstate.WorldStateStorage;
+import org.hyperledger.besu.plugin.services.storage.KeyValueStorage;
+
+import java.util.Optional;
+import java.util.function.Function;
+
+import kotlin.Pair;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import org.apache.tuweni.rlp.RLP;
+
+public class BonsaiSnapshotWorldStateKeyValueStorage extends BonsaiWorldStateKeyValueStorage
+    implements WorldStateStorage {
+
+  private final Hash worldStateRootHash;
+
+  public BonsaiSnapshotWorldStateKeyValueStorage(
+      final Hash worldStateRootHash, final StorageProvider provider) {
+    super(provider);
+    this.worldStateRootHash = worldStateRootHash;
+  }
+
+  public BonsaiSnapshotWorldStateKeyValueStorage(
+      final Hash worldStateRootHash,
+      final KeyValueStorage accountStorage,
+      final KeyValueStorage codeStorage,
+      final KeyValueStorage storageStorage,
+      final KeyValueStorage trieBranchStorage,
+      final Pair<KeyValueStorage, KeyValueStorage> snapshotTrieBranchStorage,
+      final KeyValueStorage trieLogStorage) {
+    super(
+        accountStorage,
+        codeStorage,
+        storageStorage,
+        trieBranchStorage,
+        snapshotTrieBranchStorage,
+        trieLogStorage);
+    this.worldStateRootHash = worldStateRootHash;
+  }
+
+  @Override
+  public Optional<Bytes> getAccount(final Hash accountHash) {
+    return new StoredMerklePatriciaTrie<>(
+            new StoredNodeFactory<>(
+                this::getAccountStateTrieNode, Function.identity(), Function.identity()),
+            worldStateRootHash)
+        .get(accountHash);
+  }
+
+  @Override
+  public Optional<Bytes> getStorageValueBySlotHash(final Hash accountHash, final Hash slotHash) {
+    // after a snapsync/fastsync we only have the trie branches.
+    final Optional<Bytes> account = getAccount(accountHash);
+    if (account.isPresent()) {
+      final StateTrieAccountValue accountValue =
+          StateTrieAccountValue.readFrom(
+              org.hyperledger.besu.ethereum.rlp.RLP.input(account.get()));
+      return new StoredMerklePatriciaTrie<>(
+              new StoredNodeFactory<>(
+                  (location, hash) -> getAccountStorageTrieNode(accountHash, location, hash),
+                  Function.identity(),
+                  Function.identity()),
+              accountValue.getStorageRoot())
+          .get(slotHash)
+          .map(bytes -> Bytes32.leftPad(RLP.decodeValue(bytes)));
+    }
+    return Optional.empty();
+  }
+
+  @Override
+  public Optional<Bytes> getAccountStateTrieNode(final Bytes location, final Bytes32 nodeHash) {
+    if (nodeHash.equals(MerklePatriciaTrie.EMPTY_TRIE_NODE_HASH)) {
+      return Optional.of(MerklePatriciaTrie.EMPTY_TRIE_NODE);
+    } else {
+      return snapshotTrieBranchStorage
```

Comment:
we should be able to use the blocknumber % bucket size ternary here rather than just searching both?

Reply:
I don't know if we can get this information when we call this method. Because it is called in the trie class but can be . Need to check